### PR TITLE
Several fixes around POSIX_SOURCE, remove global definitions of POSIX_SOURCE

### LINKED ISSUE #66921: arch: posix: _POSIX_C_SOURCE declaration conflicts with the rest of the system
**Describe the bug**
Picolibc and arch/posix both globally define a number of symbols reserved by IEEE-1003.1-2017 (and many earlier versions of the standard) such as `_POSIX_C_SOURCE`, `_XOPEN_SOURCE` for all Zephyr builds on that architecture, even Zephyr builds which do not explicitly want or use `CONFIG_POSIX_API=y` or any POSIX [subprofiling options](https://docs.zephyrproject.org/latest/services/portability/posix/option_groups/index.html).

This is a bit of an issue, because 
a) it means virtually every platform is in violation of [Rules A.4 and A.5 of the Coding Guidelines](https://github.com/zephyrproject-rtos/zephyr/issues/66921) since
  * ~2018 for arch/posix
  * ~2020 (based on commit date) for lib/libc/picolibc
b) since the compile definitions are hard-coded and global, they prevent applications from defining (correct, standard) application conformance levels without risking a compiler error

Screenshot 2023-12-27 at 8 49 04 PM

Screenshot 2023-12-27 at 9 22 21 PM

It seems to only be an issue since `native_sim` has facilitated running applications that enable features defined by Zephyr's POSIX API for approximately 6 months.

Please also mention any information which could help others to understand
the problem you're facing:
 - What target platform are you using? all
 - What have you tried to diagnose or workaround this issue? PRs, etc
 - Is this a regression? No

**To Reproduce**

Attempt to build any application conforming to the IEEE 1003.1-1-2017 standard with either up-to-date `_POSIX_C_SOURCE` (200809L) or up-to-date `_XOPEN_SOURCE` (700). (note: 
1.
```shell
echo -e "#define _POSIX_C_SOURCE 200809L\n#define _XOPEN_SOURCE 700" \
  | cat - samples/hello_world/src/main.c > /tmp/foo.c
mv /tmp/foo.c samples/hello_world/src/main.c
```
2. `west build -b native_sim_64 samples/hello_world`
2. Observe errors.

**Expected behavior**
Ideally the internal compiler definitions used for `ARCH_POSIX` and Picolibc would not spill out and affect the rest of the build. 

POSIX application conformance macros should only be local compile definitions, and not global compile definitions.

**Impact**
Originally this had not been noticed until after `native_sim` was added, since it was not supported to enable `CONFIG_POSIX_API` and `CONFIG_ARCH_POSIX` together for many years. They were mutually exclusive due to architectural issues.

However, now that POSIX_API is ramping up significantly and nearing feature-completion, we are seeing more frequent conflicts. Originally, this was only an issue on `native_sim` and `native_sim_64`, but now seems to be affecting all platforms, since it is a bit of a showstopper for the [POSIX Roadmap for LTSv3](https://github.com/zephyrproject-rtos/zephyr/issues/51211).

**Logs and console output**
```shell
src/main.c.obj -MF CMakeFiles/app.dir/src/main.c.obj.d -o CMakeFiles/app.dir/src/main.c.obj -c /tmp/zephyrproject/zephyr/samples/hello_world/src/main.c
/tmp/zephyrproject/zephyr/samples/hello_world/src/main.c:1: error: "_POSIX_C_SOURCE" redefined [-Werror]
    1 | #define _POSIX_C_SOURCE 200809L
      | 
<command-line>: note: this is the location of the previous definition
/tmp/zephyrproject/zephyr/samples/hello_world/src/main.c:2: error: "_XOPEN_SOURCE" redefined [-Werror]
    2 | #define _XOPEN_SOURCE 700
```

**Environment (please complete the following information):**
 - OS: Linux
 - Toolchain: Zephyr SDK v0.16.4
 - Commit: 87d056bd31b14b680e6f011134c7ce433c4462f6

**Additional context**
Related PRs (IIRC, Newlib and Picolibc might also have issues here)
https://github.com/zephyrproject-rtos/zephyr/pull/67040

### PR DESCRIPTION
**Pull request by aescolar**
Some cleanup around the selection of the API provided by the C library.
The main idea is to limit where the SOURCE selection is applied to much fewer files than before.

This PR includes some commits from #67040 so as to provide a complete solution. But it only includes those commits that are inline with the overall direction of this PR of removing all global definitions of these macros.

Background on the PR strategy differences: 
Currently #67040 cleans up only the global definitions done by picolibc, but sets newlibc to define those globally.
The current assumption in 67040 is that all functions listed in [the coding guidelines rules A.4 and A.5](https://docs.zephyrproject.org/latest/contribute/coding_guidelines/index.html#rule-a-5-c-standard-library-usage-restrictions-in-zephyr-codebase) should be globally available, without needing to set any macro by the user file, even that some of those functions are not part of the C std library API but extensions. Unfortunately, this sets a requirement on the libC integration to ensure they are exposed. See https://github.com/zephyrproject-rtos/zephyr/pull/67040#issuecomment-1896476000.

Instead, this PR solves the global use of POSIX_SOURCE issue, while postponing deciding on and solving the libC integration topic ( https://github.com/zephyrproject-rtos/zephyr/pull/67040#discussion_r1462390520 ) 

-------

Notes about the checkpatch warnings:
* For the native port specific driver "bottoms", coding guidelines rules A.4/5 do not really apply as they intend to limit the C library use in the embedded code, while this is host code. So I would think those warnings should just be ignored. A PR to clarify the rule has been submitted ( #68004 )
* The warnings in subsys/shell/modules/devmem_service.c & subsys/net/lib/lwm2m/lwm2m_rw_cbor.c are violations by the original code, which are uncovered now.
  * #68002
  * #68003
* The warning in tests/bluetooth/uuid/src/test_bt_uuid_to_str.c & tests/benchmarks/footprints/src/libc.c are due to strnlen which is an exception to rule A.4, and therefore I think the checkpatch warning in this PR should be waived.

Sidenote: We can avoid all those check-patch warnings by defining these macros in cmake, but it does not feel nice to do so, just because of checkpatch.

----

Fixes #66921

### CODE REVIEW

File: drivers/can/can_native_linux_adapt.c
Comment (keith-packard) [inline drivers/can/can_native_linux_adapt.c:116]: This change seems unrelated to the general topic, and also seems wrong -- the underlying issue is the incorrect cast of the ssize_t return from recvmsg to int.

File: drivers/eeprom/eeprom_simulator.c
Comment (keith-packard) [inline drivers/eeprom/eeprom_simulator.c:11]: Why the undef here?

**Review by keith-packard — COMMENTED** [review]
The ssize_t -> int changes are unrelated and also seem wrong? I'd also remove the `#undef` bits; they shouldn't be necessary once we get rid of the global definitions

File: drivers/can/can_native_linux_adapt.c
Comment (aescolar) [reply drivers/can/can_native_linux_adapt.c:116]: They are related in the sense that ssize_t is an extension. So either ssize_t is not used, or this files set POSIX_C_SOURCE too (note the proto is a header used by "the embedded side").
It is correct that recvmsg (or write or read) can in theory return something in the whole range of ssize_t, but given the use, the maximum they will return is the size of the buffer sent or received which cannot be bigger than the packet size for the protocols which I do not know from the top of my head, but I'm guessing not bigger than IP's 64KB, so the int range should be more than sufficient.
That's why I just took the easier approach of simple having these return int (as the users anyhow also just cast to int), instead of adding POSIX_C_SOURCE to these embedded side of these drivers.

Does that sound sensible enough?

Comment (aescolar) [reply drivers/can/can_native_linux_adapt.c:116]: Another thing I forgot yesterday, these functions are built with one C library and called from another.
ssize_t being a C library defined type can in principle not match between both, so that would also speak for only using basic types defined by the compiler.
CC: @jukkar @henrikbrixandersen as authors of this code. Your input is also welcome :)

**Review by bjarki-andreasen — DISMISSED** [review]

**Review by bjarki-andreasen — COMMENTED** [review]
I agree with the changes :)

File: drivers/can/can_native_linux_adapt.c
Comment (henrikbrixandersen) [reply drivers/can/can_native_linux_adapt.c:116]: I am fine with changing the function signature as this indeed could cause issues. To make the intention clear, perhaps add an explicit cast to int for the return value of the`recvmsg()` call?

**Review by cfriedt — CHANGES_REQUESTED** [review]
Temporary nak until the issue described in #67181 can be brought to the AWG.

**Review by cfriedt — COMMENTED** [review]
Thanks @aescolar - just a few suggestions / questions.

Currently I'm at 789/1242 tests complete and it looks reasonable.

I think this should unblock things though, at least in terms of global defines.

File: boards/posix/native_posix/CMakeLists.txt
Comment (aescolar) [reply boards/posix/native_posix/CMakeLists.txt]: changed

File: subsys/net/lib/lwm2m/lwm2m_rw_cbor.c
Comment (aescolar) [reply subsys/net/lib/lwm2m/lwm2m_rw_cbor.c:11]: I would very much have that handled in a separate discussion as to allow this fix to progress.

File: tests/lib/c_lib/strerror/src/main.c
Comment (aescolar) [reply tests/lib/c_lib/strerror/src/main.c]: Given that the test, issue and workaround existed before this PR, I think it would be best to have this discussion happen outside of this PR and any possible fix be done in another context.

File: tests/bluetooth/uuid/src/test_bt_uuid_to_str.c
Comment (aescolar) [reply tests/bluetooth/uuid/src/test_bt_uuid_to_str.c:8]: This proposal enters into the discussion from https://github.com/zephyrproject-rtos/zephyr/pull/67040#discussion_r1462390520 . Given that the current fix as is has no immediate drawbacks, and for the sake or allowing this PR to progress I would rather not re-start that discussion here again.

File: tests/lib/c_lib/strerror/src/main.c
Comment (aescolar) [reply tests/lib/c_lib/strerror/src/main.c]: Same as https://github.com/zephyrproject-rtos/zephyr/pull/67052#discussion_r1466024433

File: tests/lib/c_lib/common/src/main.c
Comment (aescolar) [reply tests/lib/c_lib/common/src/main.c:24]: Same as https://github.com/zephyrproject-rtos/zephyr/pull/67052#discussion_r1466024433

File: subsys/shell/modules/devmem_service.c
Comment (aescolar) [reply subsys/shell/modules/devmem_service.c:9]: > I think all of this can be reverted.

Each of this type of change was done to avoid a test failure

The answer from https://github.com/zephyrproject-rtos/zephyr/pull/67052#discussion_r1466024433 applies
It was agreed in the AWG to keep that RFC proposal and its discussion decoupled from this fix. Therefore I would request we do so.

File: tests/drivers/rtc/rtc_api/CMakeLists.txt
Comment (aescolar) [reply tests/drivers/rtc/rtc_api/CMakeLists.txt:38]: This is a test.
It was agreed in the AWG to keep that RFC proposal and its discussion decoupled from this fix. Therefore I would request we do so.

File: tests/net/lib/coap_client/CMakeLists.txt
Comment (aescolar) [reply tests/net/lib/coap_client/CMakeLists.txt:19]: Same answer as https://github.com/zephyrproject-rtos/zephyr/pull/67052#discussion_r1466032573
+
This test is specific for the native targets therefore, and setting the flag in general is harmless anyhow, so any filtering by target here would seem superfluous.

File: tests/subsys/modem/backends/tty/CMakeLists.txt
Comment (aescolar) [reply tests/subsys/modem/backends/tty/CMakeLists.txt:9]: This test and the module being tested are very specific for the native targets therefore, and setting the flag in general is harmless anyhow, so any filtering by target here would seem superfluous.

**Comment by aescolar** [comment]
@cfriedt The comments that seemed to require fixing have been fixed. Please note that you are still rejecting/blocking this PR even that it is fixing a bug you reported.
Given that your remaining comments seem either suggestions or discussions about future changes, or enter into other currently ongoing discussions but which are not strictly necessary to do this fix, I would suggest the rejection to be removed.

File: subsys/shell/modules/devmem_service.c
Comment (cfriedt) [reply subsys/shell/modules/devmem_service.c:9]: Yes, my original change request still applies - I guess remove the conditional I suggested if that makes you happier?

File: tests/benchmarks/footprints/src/libc.c
Comment (cfriedt) [reply tests/benchmarks/footprints/src/libc.c:11]: Please make the requested change. Only newlib  and the host libc require these application conformance macros, and we really do not need the documentation repeated here.

File: subsys/shell/modules/devmem_service.c
Comment (aescolar) [reply subsys/shell/modules/devmem_service.c:9]: Is your request that the mention of the use of getopt and the note and link be removed?

File: tests/bluetooth/uuid/src/test_bt_uuid_to_str.c
Comment (cfriedt) [reply tests/bluetooth/uuid/src/test_bt_uuid_to_str.c:8]: Please make the requested change. Only newlib  and the host libc require these application conformance macros, and we really do not need the documentation repeated here.

File: tests/lib/c_lib/common/src/main.c
Comment (cfriedt) [reply tests/lib/c_lib/common/src/main.c:24]: Please make the requested change. Only newlib  and the host libc require these application conformance macros, and we really do not need the documentation repeated here.

File: tests/bluetooth/uuid/src/test_bt_uuid_to_str.c
Comment (aescolar) [reply tests/bluetooth/uuid/src/test_bt_uuid_to_str.c:8]: In general it is not "only newlibc and the host libc" that need it, as much as "any C library which is not minimal or picolibc".
Why do you request this ifdefs be added?

Comment (aescolar) [reply tests/bluetooth/uuid/src/test_bt_uuid_to_str.c:8]: @cfriedt Your original proposal has several changes:
a) You use CONFIG_HOST_LIBC but that does not exist I presume if anything that would  need to be EXTERNAL_LIBC
b) Please elaborate on why you want to `#if` these defines and if you will block this PR unless a similar `#if defined(..`  is done
c) Please clarify if you also will block this PR unless the comment is removed.

**Comment by cfriedt** [comment]
> @cfriedt The comments that seemed to require fixing have been fixed. Please note that you are still rejecting/blocking this PR even that it is fixing a bug you reported.

Yes, I would still like to see some requested changes made.

Also, when you updated your code it was 3am for me. I sleep (sometimes).

> Given that your remaining comments seem either suggestions or discussions about future changes, or enter into other currently ongoing discussions but which are not strictly necessary to do this fix, I would suggest the rejection to be removed.

I've resolved a few comments. There are still some that I would like to see changed.

**Comment by aescolar** [comment]
> There are still some that I would like to see changed.

@cfriedt Would you be so kind as to respond to the questions/clarification requests in your comments? That will probably speed up things.

**Comment by cfriedt** [comment]
> > There are still some that I would like to see changed.
> 
> @cfriedt Would you be so kind as to respond to the questions/clarification requests in your comments? That will probably speed up things.

I've been doing that since 5:55am 😁 You do realize I'm not in your timezone, right?

File: subsys/shell/modules/devmem_service.c
Comment (cfriedt) [reply subsys/shell/modules/devmem_service.c:9]: The comment is not necessary or necessarily accurate. `CONFIG_GETOPT` is used in several places within Zephyr.

Please remove it.

File: tests/bluetooth/uuid/src/test_bt_uuid_to_str.c
Comment (cfriedt) [reply tests/bluetooth/uuid/src/test_bt_uuid_to_str.c:8]: > In general it is not "only newlibc and the host libc" that need it, as much as "any C library which is not minimal or picolibc".

Actually, there are certified ISO C libraries that do not provide the (POSIX) `strnlen()`.

>  Why do you request this ifdefs be added?

Because currently only the host libc and newlib require them. I believe this is self-explanatory. 

> The reason to mention explicitly which functions require this

I understand. However, it's covered in the documentation already. Having a single source of truth is nice (i.e. the documentation).

> is because we are tripping checkpatch

The solution in this PR does trip checkpatch.pl - there are alternatives that are probably more scalable and maintainable that do not trip checkpatch.pl, that do not involve hard-coding values in source files.. but that's a separate topic.

>  Please clarify if you also will block this PR unless the comment is removed.

I would prefer if the comment were removed.

Comment (aescolar) [reply tests/bluetooth/uuid/src/test_bt_uuid_to_str.c:8]: >> Why do you request this ifdefs be added?

> Because currently only the host libc and newlib require them. I believe this is self-explanatory.

Unfortunately it is not self explanatory. You are requesting adding extra code to set conditionally a flag which can be set always. And I would prefer to not add unnecessary code.
So I would request you to reconsider rejecting this PR based on that.

File: tests/lib/c_lib/strerror/src/main.c
Comment (aescolar) [reply tests/lib/c_lib/strerror/src/main.c]: The test was removed in https://github.com/zephyrproject-rtos/zephyr/pull/67891 . Therefore setting this macro is not necessary anymore.

**Review by henrikbrixandersen — DISMISSED** [review]

**Comment by henrikbrixandersen** [comment]
The compliance checks are false positives.

**Review by henrikbrixandersen — DISMISSED** [review]

**Review by jukkar — APPROVED** [review]

**Review by henrikbrixandersen — APPROVED** [review]

**Review by bjarki-andreasen — APPROVED** [review]

**Review by de-nordic — APPROVED** [review]
Looks OK. Thanks!

**Review by rlubos — APPROVED** [review]
Net part looks ok

**Review by cfriedt — APPROVED** [review]
LGTM - thanks @aescolar 👍

**Pull request merged**
